File: lorcast.py
"""
lorcast.py — Thin client for the Lorcast public API (https://lorcast.com).

Endpoint used: GET /cards/:set/:number  (one set+number → one card object)

A persistent on-disk JSON cache means each card is fetched at most once.
Images (AVIF) are also cached on disk under cache/images/.

If the network is unavailable or a card cannot be found, lookups return
None and the UI falls back to the data already extracted from the PDF.
"""

from __future__ import annotations

import json
import time
import urllib.request
import urllib.error
from dataclasses import dataclass
from pathlib import Path

API_BASE = "https://api.lorcast.com/v0"
USER_AGENT = "LorcanaPickingTool/1.0 (+local)"
REQUEST_TIMEOUT = 15  # seconds


@dataclass
class LorcastCard:
    """Subset of the Lorcast card model that we need."""
    name: str
    version: str | None
    ink: str | None              # "Amber", "Amethyst", ...
    rarity: str | None           # "Common", "Rare", ...
    set_code: str
    set_name: str | None
    collector_number: str
    image_small: str | None      # url
    image_normal: str | None     # url


class LorcastClient:
    def __init__(self, cache_dir: str | Path):
        self.cache_dir = Path(cache_dir)
        self.cache_dir.mkdir(parents=True, exist_ok=True)
        self.image_dir = self.cache_dir / "images"
        self.image_dir.mkdir(parents=True, exist_ok=True)
        self.card_cache_file = self.cache_dir / "cards.json"
        self._card_cache: dict[str, dict] = {}
        if self.card_cache_file.exists():
            try:
                self._card_cache = json.loads(self.card_cache_file.read_text("utf-8"))
            except json.JSONDecodeError:
                self._card_cache = {}
        # remember misses too, with a timestamp so we can re-try later
        self._miss_cache: dict[str, float] = {}

    # ---- public lookups -----------------------------------------------------

    def get_card(self, set_code: str, number: str) -> LorcastCard | None:
        """Return the card or None on miss/network error."""
        key = f"{set_code}/{number}"
        # cache hit?
        if key in self._card_cache:
            return _card_from_dict(self._card_cache[key])
        # recent miss? (don't hammer the API for cards it doesn't know)
        last_miss = self._miss_cache.get(key)
        if last_miss and (time.time() - last_miss) < 3600:
            return None

        url = f"{API_BASE}/cards/{set_code}/{number}"
        try:
            data = self._http_get_json(url)
        except (urllib.error.HTTPError, urllib.error.URLError, TimeoutError):
            self._miss_cache[key] = time.time()
            return None

        self._card_cache[key] = data
        self._save_cache()
        return _card_from_dict(data)

    def get_image_path(self, card: LorcastCard, size: str = "small") -> Path | None:
        """Download (and cache) the card image, return its local path."""
        url = card.image_small if size == "small" else card.image_normal
        if not url:
            return None
        # filename = set + number + size + extension
        ext = url.rsplit("?", 1)[0].rsplit(".", 1)[-1].lower()
        if ext not in ("avif", "png", "jpg", "jpeg", "webp"):
            ext = "avif"
        fname = f"{card.set_code}_{card.collector_number}_{size}.{ext}"
        local = self.image_dir / fname
        if local.exists() and local.stat().st_size > 0:
            return local

        try:
            req = urllib.request.Request(url, headers={"User-Agent": USER_AGENT})
            with urllib.request.urlopen(req, timeout=REQUEST_TIMEOUT) as r:
                local.write_bytes(r.read())
        except (urllib.error.HTTPError, urllib.error.URLError, TimeoutError):
            return None
        return local

    # ---- internals ----------------------------------------------------------

    def _http_get_json(self, url: str) -> dict:
        req = urllib.request.Request(
            url,
            headers={"User-Agent": USER_AGENT, "Accept": "application/json"},
        )
        with urllib.request.urlopen(req, timeout=REQUEST_TIMEOUT) as r:
            return json.loads(r.read())

    def _save_cache(self) -> None:
        tmp = self.card_cache_file.with_suffix(".tmp")
        tmp.write_text(json.dumps(self._card_cache, ensure_ascii=False), "utf-8")
        tmp.replace(self.card_cache_file)


def _card_from_dict(d: dict) -> LorcastCard:
    img = (d.get("image_uris") or {}).get("digital") or {}
    s = d.get("set") or {}
    return LorcastCard(
        name=d.get("name", ""),
        version=d.get("version"),
        ink=d.get("ink"),
        rarity=d.get("rarity"),
        set_code=str(s.get("code", "")),
        set_name=s.get("name"),
        collector_number=str(d.get("collector_number", "")),
        image_small=img.get("small"),
        image_normal=img.get("normal"),
    )

File: src/main/lorcast.ts
// Client Lorcast (port du lorcast.py V1) : une carte n'est demandée qu'une
// seule fois à l'API, tout est mis en cache sur disque (JSON + images AVIF).
// Hors-ligne ou carte inconnue → null, l'app retombe sur les données du PDF.

import { app } from 'electron'
import { existsSync, mkdirSync, readFileSync, statSync, writeFileSync, renameSync } from 'fs'
import { writeFile } from 'fs/promises'
import { join } from 'path'

const API_BASE = 'https://api.lorcast.com/v0'
const USER_AGENT = 'LorcanaPickingTool/2.0 (+local)'
const MISS_TTL_MS = 3600_000 // ne pas marteler l'API pour une carte inconnue

export interface LorcastCard {
  name: string
  version: string | null
  ink: string | null // "Amber", ...
  rarity: string | null // "Common", "Super_rare", ...
  set_code: string
  set_name: string | null
  collector_number: string
  image_small: string | null
  image_normal: string | null
  /** vignette dans le cache images, si téléchargée */
  image_file: string | null
  /** version haute définition (zoom), si téléchargée */
  image_large_file: string | null
}

let cardCache: Record<string, unknown> | null = null
const missCache = new Map<string, number>()

function cacheDir(): string {
  const dir = join(app.getPath('userData'), 'cache')
  mkdirSync(join(dir, 'images'), { recursive: true })
  return dir
}

function cacheFile(): string {
  return join(cacheDir(), 'cards.json')
}

function loadCache(): Record<string, unknown> {
  if (cardCache) return cardCache
  try {
    cardCache = JSON.parse(readFileSync(cacheFile(), 'utf-8'))
  } catch {
    cardCache = {}
  }
  return cardCache!
}

function saveCache(): void {
  const tmp = cacheFile() + '.tmp'
  writeFileSync(tmp, JSON.stringify(cardCache ?? {}), 'utf-8')
  renameSync(tmp, cacheFile())
}

function fromApi(d: Record<string, unknown>): Omit<LorcastCard, 'image_file' | 'image_large_file'> {
  const img = ((d.image_uris as Record<string, unknown>)?.digital ?? {}) as Record<string, string>
  const set = (d.set ?? {}) as Record<string, unknown>
  return {
    name: (d.name as string) ?? '',
    version: (d.version as string) ?? null,
    ink: (d.ink as string) ?? null,
    rarity: (d.rarity as string) ?? null,
    set_code: String(set.code ?? ''),
    set_name: (set.name as string) ?? null,
    collector_number: String(d.collector_number ?? ''),
    image_small: img.small ?? null,
    image_normal: img.normal ?? null
  }
}

/** Carte par chapitre + numéro de collection, avec image téléchargée en cache. */
export async function getCard(setCode: string, number: string): Promise<LorcastCard | null> {
  const key = `${setCode}/${number}`
  const cache = loadCache()

  let raw = cache[key] as Record<string, unknown> | undefined
  if (!raw) {
    const lastMiss = missCache.get(key)
    if (lastMiss && Date.now() - lastMiss < MISS_TTL_MS) return null
    try {
      const res = await fetch(`${API_BASE}/cards/${setCode}/${number}`, {
        headers: { 'User-Agent': USER_AGENT, Accept: 'application/json' },
        signal: AbortSignal.timeout(15_000)
      })
      if (!res.ok) throw new Error(`HTTP ${res.status}`)
      raw = (await res.json()) as Record<string, unknown>
      cache[key] = raw
      saveCache()
    } catch {
      missCache.set(key, Date.now())
      return null
    }
  }

  const card = fromApi(raw)
  const image_file = await ensureImage(card, 'small')
  const image_large_file = await ensureImage(card, 'large')
  return { ...card, image_file, image_large_file }
}

/**
 * Télécharge (une fois) l'image de la carte, renvoie le nom de fichier local.
 * 'small' = vignette (listes), 'large' = haute définition (zoom plein écran).
 */
async function ensureImage(
  card: Omit<LorcastCard, 'image_file' | 'image_large_file'>,
  size: 'small' | 'large'
): Promise<string | null> {
  const url = size === 'small' ? (card.image_small ?? card.image_normal) : (card.image_normal ?? card.image_small)
  if (!url) return null
  let ext = url.split('?')[0].split('.').pop()?.toLowerCase() ?? 'avif'
  if (!['avif', 'png', 'jpg', 'jpeg', 'webp'].includes(ext)) ext = 'avif'
  const fname = `${card.set_code}_${card.collector_number}_${size}.${ext}`.replace(/[^\w.-]/g, '_')
  const local = join(cacheDir(), 'images', fname)
  if (existsSync(local) && statSync(local).size > 0) return fname

  try {
    const res = await fetch(url, {
      headers: { 'User-Agent': USER_AGENT },
      signal: AbortSignal.timeout(15_000)
    })
    if (!res.ok) throw new Error(`HTTP ${res.status}`)
    await writeFile(local, Buffer.from(await res.arrayBuffer()))
    return fname
  } catch {
    return null
  }
}

export function imagesDir(): string {
  return join(cacheDir(), 'images')
}
